lib/Analysis/LevelAnalysis/LevelAnalysis.cpp

```diff
@@ -50,6 +57,16 @@ LogicalResult LevelAnalysis::visitOperation(
         auto level = operandLattice->getValue().getLevel();
         propagate(modReduceOp.getResult(), LevelState(level + 1));
       })
+      .Case<mgmt::LevelReduceOp>([&](auto levelReduceOp) {
+        // implicitly ensure that the operand is secret
+        const auto *operandLattice = operands[0];
+        if (!operandLattice->getValue().isInitialized()) {
+          return;
+        }
+        auto level = operandLattice->getValue().getLevel();
+        propagate(levelReduceOp.getResult(),
+                  LevelState(level + levelReduceOp.getLevelToDrop()));
```

Comment:
I would expect this to be a subtraction, if the input starts at level `L` and goes down to 0. Though some of your tests have this increasing when a level reduce op is applied.

Similar below where you lattice join with a max instead of a min (though maybe that makes sense because you're backpropagating?)

Reply:
See comment in `LevelAnalysis` docstring. Inside level analysis we count from 0 to L instead of L to 0 as we does not know L in the first place.